Add individual analytics leaderboard endpoints

joshbetz opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Register routes for each of the individual leaderboards.

These leaderboards can be pretty expensive to generate. If you only need
to data from one of them, it can be a pretty significant performance
boost to avoid generating all the leaderboards.

### How to test the changes in this Pull Request:

1. GET /wc-analytics/leaderboards. Ensure all 4 groups (products, customers, coupons, categories) are in the response.
2. GET /wc-analytics/leaderboards/products. Ensure only the products group is in the response.
3. GET /wc-analytics/leaderboards/customers. Ensure only the customers group is in the response.
4. GET /wc-analytics/leaderboards/coupons. Ensure only the coupons group is in the response.
5. GET /wc-analytics/leaderboards/categories. Ensure only the categories group is in the response.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshbetz committed:
Add individual analytics leaderboard endpoints

Register routes for each of the individual leaderboards.

These leaderboards can be pretty expensive to generate. If you only need
to data from one of them, it can be a pretty significant performance
boost to avoid generating all the leaderboards.

botwoo commented:
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33185/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `Add individual analytics leaderboard endpoints` 80a7ab332dbe8a600a2d6b9dd2948984d88ee850
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

joshbetz committed:
lint

joshbetz committed:
changelog

coreymckrill commented on plugins/woocommerce/src/Admin/API/Leaderboards.php:
```diff
@@ -53,6 +53,22 @@ public function register_routes() {
 			)
 		);
 
+		foreach ( [ 'customers', 'coupons', 'categories', 'products' ] as $endpoint ) {
+			register_rest_route(
+				$this->namespace,
+				'/' . $this->rest_base . '/' . $endpoint,
+				array(
+					array(
+						'methods'             => \WP_REST_Server::READABLE,
+						'callback'            => array( $this, 'get_items' ),
+						'permission_callback' => array( $this, 'get_items_permissions_check' ),
+						'args'                => $this->get_collection_params(),
+					),
+					'schema' => array( $this, 'get_public_item_schema' ),
+				)
+			);
+		}
```
Instead of creating four separate endpoints, one alternative would be to use a URI argument instead. Something like this:

```php
		register_rest_route(
				$this->namespace,
				'/' . $this->rest_base . '/(?P<leaderboard>[\w-]+)',
				array(
					'args' => array(
						'leaderboard' => array(
							'type' => 'string',
							'enum' => array( 'customers', 'coupons', 'categories', 'products' ),
						),
					)
					array(
						'methods'             => \WP_REST_Server::READABLE,
						'callback'            => array( $this, 'get_items' ),
						'permission_callback' => array( $this, 'get_items_permissions_check' ),
						'args'                => $this->get_collection_params(),
					),
					'schema' => array( $this, 'get_public_item_schema' ),
				)
			);
```

I'm not sure this has any huge advantages over the other approach, but it does simplify things a bit by having less total endpoints, and gives you built-in validation and error handling for the endpoint URI, which seems a little less fragile than traversing to the end of an array to determine which leaderboard to show.

coreymckrill commented on plugins/woocommerce/src/Admin/API/Leaderboards.php:
```diff
@@ -369,9 +385,29 @@ public function get_leaderboards( $per_page, $after, $before, $persisted_query )
+			case 'products':
+				$leaderboards = array( $this->get_products_leaderboard( $request['per_page'], $request['after'], $request['before'], $persisted_query ) );
+				break;
+		}
```
This should probably either define `$leaderboards` as an empty array before the switch or have a `default` case, to avoid notices about undefined variables.

joshbetz committed:
Use a URI argument instead of 4 separate endpoints

This feels nicer for registering the endpoint and we also don't have to
manually parse the URL to figure out which report is needed.

props @coreymckrill

coreymckrill reviewed (dismissed):
👍 Working well on local

chihsuan reviewed (changes requested):
Thanks @joshbetz. 

I noticed that unit test CI is failing. Could you take a look? I think this affects the original behavior of `/leaderboards` API. We should update the unit tests to reflect the changes.


```
There were 2 failures:

***) WC_Admin_Tests_API_Leaderboards::test_schema
Failed asserting that actual size 4 matches expected size 3.

/home/runner/work/woocommerce/woocommerce/plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/api/leaderboards.php:67
phpvfscomposer:///home/runner/work/woocommerce/woocommerce/plugins/woocommerce/vendor/phpunit/phpunit/phpunit:60

2) WC_Admin_Tests_API_Leaderboards::test_filter_leaderboards
Failed asserting that 400 matches expected 200.
```

joshbetz committed:
Register /allowed before individual leaderboards

If /allowed is registered last, it doesn't work because the validation
on the individual leaderboards doesn't expect `allowed` to be an option.

joshbetz commented:
Thanks @chihsuan. That's fix now.

chihsuan reviewed (commented):
Thanks @joshbetz Looks good and tested well! Could you also please add some tests for new endpoints?

https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/API/Leaderboards.php

joshbetz committed:
Add tests for individual leaderboard endpoints

Ensures each endpoint returns the correct group

chihsuan reviewed (approved):
Well done! @joshbetz 🚀
